src/dune_pkg/lock_dir.ml

```diff
@@ -1389,11 +1389,44 @@ module Write_disk = struct
     | Error e -> raise_user_error_on_check_existance path e
   ;;
 
+  let rec safely_copy_lock_dir_when_dst_non_existant ~dst src =
+    match check_existing_lock_dir src with
+    | Ok `Is_existing_lock_dir ->
+      (match Path.readdir_unsorted_with_kinds src with
+       | Error (e, _, _) ->
+         User_error.raise
+           [ Pp.textf
+               "Failed to list %s with error %s"
+               (Path.to_string_maybe_quoted src)
+               (Unix.error_message e)
+           ]
+       | Ok children ->
+         List.iter
+           ~f:(fun (name, kind) ->
+             let child_src = Path.relative src name in
+             let child_dst = Path.relative dst name in
+             match kind with
+             | Unix.S_DIR ->
+               Path.mkdir_p child_dst;
+               safely_copy_lock_dir_when_dst_non_existant ~dst:child_dst child_src
+             | _ ->
+               Path.mkdir_p dst;
+               Io.copy_file ~src:child_src ~dst:child_dst ())
+           children)
+    | Ok `Non_existant -> ()
+    | Error e -> raise_user_error_on_check_existance src e
+  ;;
+
   (* Does the same checks as [safely_remove_lock_dir_if_exists_thunk] but it raises an
      error if the lock dir already exists. [dst] is the new file name *)
   let safely_rename_lock_dir_thunk ~dst src =
     match check_existing_lock_dir src, check_existing_lock_dir dst with
-    | Ok `Is_existing_lock_dir, Ok `Non_existant -> fun () -> Path.rename src dst
+    | Ok `Is_existing_lock_dir, Ok `Non_existant ->
+      fun () ->
+        (try Path.rename src dst with
```

Comment:
Can you switch this to `match .. with | exception ...`?

Reply:
Not meaning to block your review request, but why is that better *in this case*?